src/analysis.py

```diff
@@ -0,0 +1,147 @@
+import numpy as np
+import scipy.signal as sig
+import math
+import statistics
+
+
+def calculate_grating_frequencies(grating, distanceperpixel):
+    periods = []
+    frequencies = []
+    for row in grating:
+        grating_frequencies, grating_periods = calculate_row_frequencies(
+            row=row,
+            number_of_frequencies=3,
+            micrometers_per_pixel=distanceperpixel)
+        periods.append(grating_periods)
+        frequencies.append(grating_frequencies)
+    periods = np.array(periods).T
+    frequencies = np.array(frequencies).T
+
+    period_average = [np.sum(p) / len(p) for p in periods]
+    period_errors = [StandardErrorMean(x=p) for p in periods]
+    frequency_average = [np.sum(f) / len(f) for f in frequencies]
+    frequency_errors = [StandardErrorMean(x=f) for f in frequencies]
+
+    return {
+        'Average_Periods_nm': period_average,
+        'Period_Errors_nm': period_errors,
+        'Average_Frequencies': frequency_average,
+        'Frequencies_Errors': frequency_errors}
+
+
+def calculate_row_frequencies(row,
+                          number_of_frequencies,
+                          micrometers_per_pixel):
+    '''
+    Calculate fourier transform of image row.
+    Args:
+        row: <array> pixel values for image row
+        number_of_frequencies: <int> number of peaks to pull from fourier transform (number
+                    of periods to analyse)
+        micrometers_per_pixel: <float> distance in um per pixel
+    Returns:
+        frequencies: <array> fourier space frequency values of period peaks
+        periods: <array> signal periods from fourier transform converted from um
+                    to nm
+    '''
+    frequency_intensities = np.fft.rfft(row)
+    absolute_frequency_intensities = np.abs(frequency_intensities)
+
+    peak_locations, _ = sig.find_peaks(x=absolute_frequency_intensities)
+    prominences, _, _ = sig.peak_prominences(
+        x=absolute_frequency_intensities,
+        peaks=peak_locations)
+    locations_sorted_by_prominence = np.array(
+        np.argsort(prominences)[-number_of_frequencies:][::-1])
+    selected_peak_locations = peak_locations[locations_sorted_by_prominence]
+
+    samplesize = len(row)
+    frequency_coordinates = np.fft.rfftfreq(samplesize, 1)
+    frequencies = frequency_coordinates[locations_sorted_by_prominence]
+
+    frequency_steps = [p / (micrometers_per_pixel * samplesize)
+                       for p in selected_peak_locations]
+    periods = [(1 / f) * 1E3 for f in frequency_steps]
+
```

Comment:
Is there some redundancy in lines 58 to 64? [59, 60] look quite similar to [62, 64].